psql: add tab-completion support for CREATE TABLE

stanislavlyalin opened the pull request:
Add psql tab-completion support for CREATE TABLE commands.

This patch adds tab-completion support for CREATE TABLE commands in
accordance with the Greenplum Database documentation.

The patch provides keyword completion for CREATE TABLE, data type completion,
column name completion in the appropriate contexts, and property completion
within WITH clauses.

A new helper function prev_words_match was introduced to search a list of
previous tokens for a sub-sequence of words, either consecutive or
non-consecutive, with support for wildcard placeholders.

Additionally, schema name completion was fixed for cases where only a single
schema exists (for example, "public.").

Syntax diagram:

<img width="3028" height="142" alt="image" src="https://github.com/user-attachments/assets/1a7a89f7-7b99-49d0-a672-4e57dedfd225" />

Demo:

![Peek 2025-10-07 18-28](https://github.com/user-attachments/assets/39555867-2089-405f-bf4b-7500b68ee0f9)

stanislavlyalin commented on src/bin/psql/tab-complete.c:
```diff
@@ -4244,6 +4308,152 @@ get_previous_words(int point, char **previous_words, int nwords)
+			const char *candidate_word = previous_words[word_index--];
+
+			bool matches = (candidate_word &&
+							*candidate_word != '\0' &&
```
This more clearly shows that tokens cannot be empty lines. But there is no great sense in this.

stanislavlyalin commented on src/bin/psql/tab-complete.c:
```diff
@@ -4244,6 +4308,152 @@ get_previous_words(int point, char **previous_words, int nwords)
+ * NULL/empty entries are skipped. Case-insensitive compare via pg_strcasecmp().
+ */
+static bool
+previous_words_match(char *previous_words[], int n_expected, ...)
```
I simplified the function for two words in the arguments, and also added its call for the case of `PARTITION BY`.

stanislavlyalin commented on src/bin/psql/tab-complete.c:
```diff
@@ -2271,6 +2281,60 @@ psql_completion(const char *text, int start, int end)
+			COMPLETE_WITH_LIST(cols);
+
+		for (int i = 0; i < ncols; ++i) free(cols[i]);
+		free(cols);
```
If no columns are found, the `free(cols);` call will be made on a null pointer. According to https://en.cppreference.com/w/c/memory/free, calling `free(NULL);` is safe.

andr-sokolov commented on src/bin/psql/tab-complete.c:
```diff
@@ -840,6 +847,9 @@ initialize_readline(void)
 
 
+/* Max number of previous words scanned for completion context. */
+enum { LOOKBACK_LIMIT = 20 };
```
How this number has been calculated?

stanislavlyalin replied:
According to [CREATE TABLE definition](https://techdocs.broadcom.com/us/en/vmware-tanzu/data-solutions/tanzu-greenplum/6/greenplum-database/ref_guide-sql_commands-CREATE_TABLE.html) maximum number of keywords is **24**:
```sql
1 CREATE
2 GLOBAL
3 TEMP
4 TABLE
5 IF
6 NOT
7 EXISTS
8 TableName
9 (columns)
10 INHERITS
11 (...)
12 WITH
13 (...)
14 ON
15 COMMIT
16 PRESERVE
17 ROWS
18 TABLECPASE
19 <tablespace_name>
20 DISTRIBUTED
21 BY
22 (column)
23 PARTITION
24 BY
```
So, with a margin we can take **30 elements**.

stanislavlyalin commented:
### Fixes and improvements

- Fixed schema name completion after CREATE TABLE.
- Fixed keyword completion after CREATE TABLE, restricted to cases outside of the column definition context (i.e., when there is no single bare opening parenthesis).
- Fixed bug: completion word lists were not terminated with NULL.
- Added column type completion, including experimental numeric types and keywords.
- Added completion after LOCAL|GLOBAL TEMPORARY, which was missing in the original implementation.

### Limitations

- Column name completion for PARTITION BY RANGE( is not implemented to avoid overly complex logic.
- Parameter completion inside WITH( is not implemented because it is difficult to distinguish the closing parenthesis of WITH from that of BY. In addition, WITH options can appear after an arbitrary number of tokens; binding strictly to TABLE WITH would conflict with the CREATE TABLE branch.
- If nested parentheses appear inside column definitions (e.g., char(8)), completion will stop working afterwards. I could not find a simple algorithm that would both handle this case and reliably distinguish column-definition parentheses from those used for WITH, BY, RANGE, etc.

andr-sokolov commented:
>     * Parameter completion inside WITH( is not implemented because it is difficult to distinguish the closing parenthesis of WITH from that of BY. In addition, WITH options can appear after an arbitrary number of tokens; binding strictly to TABLE WITH would conflict with the CREATE TABLE branch.


autocompletion for `with(appendonly=true);` is very useful for Greengage. What about not implementing something, but implementing autocompletion for storage options?
